application/src/main/java/org/togetherjava/tjbot/commands/filesharing/FileSharingMessageListener.java

```diff
@@ -0,0 +1,197 @@
+import org.jetbrains.annotations.NotNull;
+import org.togetherjava.tjbot.commands.MessageReceiverAdapter;
+import org.togetherjava.tjbot.config.Config;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+
+/**
+ * Listener that receives all sent help messages and uploads them to a share service if the message
+ * contains a file with the given extension in the
+ * {@link FileSharingMessageListener#extensionFilter}.
+ */
+public class FileSharingMessageListener extends MessageReceiverAdapter {
+
+    private static final ObjectMapper JSON = new ObjectMapper();
+
+    private final String gistApiKey;
+    private static final String SHARE_API = "https://api.github.com/gists";
+    private static final HttpClient CLIENT = HttpClient.newHttpClient();
+
+    private final Set<String> extensionFilter = Set.of("txt", "java", "gradle", "xml", "kt", "json",
+            "fxml", "css", "c", "h", "cpp", "py", "yml");
+
+    private final Predicate<String> isStagingChannelName;
+    private final Predicate<String> isOverviewChannelName;
+
+
+    public FileSharingMessageListener(@NotNull Config config) {
+        super(Pattern.compile(".*"));
+
+        gistApiKey = config.getGistApiKey();
+        isStagingChannelName = Pattern.compile(config.getHelpSystem().getStagingChannelPattern())
+            .asMatchPredicate();
+        isOverviewChannelName = Pattern.compile(config.getHelpSystem().getOverviewChannelPattern())
+            .asMatchPredicate();
+    }
+
+    @Override
+    public void onMessageReceived(@NotNull MessageReceivedEvent event) {
+        User author = event.getAuthor();
+        if (author.isBot() || event.isWebhookMessage()) {
+            return;
+        }
+
+        if (!isHelpThread(event)) {
+            return;
+        }
+
+
+        List<Message.Attachment> attachments = event.getMessage()
+            .getAttachments()
+            .stream()
+            .filter(this::isAttachmentRelevant)
+            .toList();
+
+        CompletableFuture.runAsync(() -> processAttachments(event, attachments));
+    }
+
+    private boolean isAttachmentRelevant(@NotNull Message.Attachment attachment) {
+        String extension = attachment.getFileExtension();
+        if (extension == null) {
+            return false;
+        }
+        return extensionFilter.contains(extension);
+    }
+
+
+    private void processAttachments(@NotNull MessageReceivedEvent event,
+            @NotNull List<Message.Attachment> attachments) {
+
+        Map<String, GistFile> filesAsJson = Collections.synchronizedMap(new HashMap<>());
```

Comment:
Why are you not using a ConcurrentHashMap?

Reply:
ask Zabu tbh, he told me to update it like this

Reply:
also does it make a difference?

Reply:
ConcurrentHashMap is to make the app thread safe. So u should use it if doing multi-threading

Reply:
Reason why I ask is because ConcurrentHashMap is faster, there's some more differences (like that Collections#syncronizedMap allows null keys, etc), but unless you need one of those I don't see why you should use it.

Reply:
but Collections.synchronizedMap returns a thread safe map says the docs